File: aurora/android/app/src/main/java/app/gamenative/utils/preInstallSteps/DirectXStep.kt
package app.gamenative.utils

import app.gamenative.data.GameSource
import app.gamenative.enums.Marker
import com.winlator.container.Container
import java.io.File
import timber.log.Timber

/**
 * Aurora Phase 7a: DirectX Runtime auto-installer.
 *
 * Scans the game directory for DirectX installers (DXSETUP.exe) and
 * queues them for silent installation. Required by most pre-2015 games.
 *
 * Aurora-specific addition (not in GameNative).
 */
object DirectXStep : PreInstallStep {
    override val marker: Marker = Marker.DIRECTX_INSTALLED

    override fun appliesTo(
        container: Container,
        gameSource: GameSource,
        gameDirPath: String,
    ): Boolean {
        return !MarkerUtils.hasMarker(gameDirPath, Marker.DIRECTX_INSTALLED)
    }

    override fun buildCommand(
        container: Container,
        appId: String,
        gameSource: GameSource,
        gameDir: File,
        gameDirPath: String,
    ): String? {
        val searchDirs = listOf(
            File(gameDirPath, "_CommonRedist/DirectX"),
            File(gameDirPath, "DirectX"),
            File(gameDirPath, "Redist/DirectX"),
            File(gameDirPath, "__Installer/directx"),
            File(gameDirPath, "__Installer/directx/redist"),
            File(gameDirPath, "redist"),
            File(gameDirPath, "Redist"),
        ).filter { it.exists() && it.isDirectory }

        if (searchDirs.isEmpty()) {
            Timber.tag("DirectXStep").i("No DirectX search directories found")
            return null
        }

        val parts = mutableListOf<String>()

        for (dir in searchDirs) {
            Timber.tag("DirectXStep").i("Searching for DXSETUP.exe under ${dir.absolutePath}")
            dir.walkTopDown()
                .filter { file ->
                    file.isFile &&
                        file.name.equals("DXSETUP.exe", ignoreCase = true)
                }
                .forEach { installerFile ->
                    val relativePath = installerFile
                        .relativeTo(gameDir)
                        .path
                        .replace('/', '\\')
                    val winePath = "A:\\$relativePath"
                    Timber.tag("DirectXStep").i("Queued DirectX installer: $winePath")
                    parts.add("$winePath /silent")
                }
        }

        return if (parts.isEmpty()) null else parts.joinToString(" & ")
    }
}

File: aurora/archive/poc_engines/src/auto_installer/steps/directx_step.py
#!/usr/bin/env python3
"""
Aurora Emulator - Phase 7: DirectX Step
=========================================

Auto-detects and installs DirectX runtime (DirectX 9.0c June 2010 redist).
Aurora-specific addition (not in GameNative).

Most pre-2015 Windows games require DirectX 9.0c runtime files
(d3dx9_*.dll, d3dcompiler_*.dll). The DirectX redist installer extracts
these to C:\\Windows\\System32 inside the Wine prefix.
"""

from __future__ import annotations

from pathlib import Path

from ..marker_utils import Marker, has_marker
from ..pre_install_step import PreInstallStep, InstallCommand


# Known DirectX installer paths (Windows paths, A:\ = game root)
DIRECTX_PATHS: list[tuple[str, str]] = [
    # Standard Steam/GOG layout
    ("A:\\_CommonRedist\\DirectX\\Jun2010\\DXSETUP.exe", "/silent"),
    ("A:\\_CommonRedist\\DirectX\\Jun2010_redist\\DXSETUP.exe", "/silent"),
    ("A:\\_CommonRedist\\DirectX\\Jun2010\\dxsetup.exe", "/silent"),
    # Older DirectX redist versions
    ("A:\\_CommonRedist\\DirectX\\Mar2009\\DXSETUP.exe", "/silent"),
    ("A:\\_CommonRedist\\DirectX\\Aug2008\\DXSETUP.exe", "/silent"),
    # Root-level DXSETUP
    ("A:\\DirectX\\DXSETUP.exe", "/silent"),
    ("A:\\redist\\DXSETUP.exe", "/silent"),
    # Some games put it in a Redist subdir
    ("A:\\Redist\\DirectX\\DXSETUP.exe", "/silent"),
    # GOG installers
    ("A:\\__Installer\\directx\\DXSETUP.exe", "/silent"),
    # EA/Origin games
    ("A:\\__Installer\\directx\\redist\\DXSETUP.exe", "/silent"),
]


class DirectXStep(PreInstallStep):
    """Auto-install DirectX 9.0c runtime."""

    @property
    def marker(self) -> Marker:
        return Marker.DIRECTX_INSTALLED

    @property
    def name(self) -> str:
        return "DirectX Runtime"

    def applies_to(self, game_dir: Path) -> bool:
        return not has_marker(game_dir, self.marker)

    def detect(self, game_dir: Path) -> list[InstallCommand]:
        commands: list[InstallCommand] = []
        for win_path, args in DIRECTX_PATHS:
            if not win_path.startswith("A:\\"):
                continue
            rest = win_path[3:]
            host_path = game_dir / rest.replace("\\", "/")
            if host_path.is_file():
                commands.append(InstallCommand(
                    executable=win_path,
                    args=args,
                    description="DirectX 9.0c Runtime",
                ))
                break  # Only need one DirectX install
        return commands
